usvm-ts/src/main/kotlin/org/usvm/machine/operator/TsBinaryOperator.kt

```diff
@@ -97,294 +99,274 @@ sealed interface TsBinaryOperator {
+        reduce: (List<ExprWithTypeConstraint<R>>) -> UExpr<R>,
+    ): UExpr<R>? = with(lhs.tctx) {
+        check(lhs.isFakeObject() || rhs.isFakeObject())
+
+        val conjuncts = mutableListOf<ExprWithTypeConstraint<R>>()
+
+        when {
+            lhs.isFakeObject() && rhs.isFakeObject() -> {
+                val lhsType = lhs.getFakeType(scope)
+                val rhsType = rhs.getFakeType(scope)
+
+                // fake(bool) + fake(bool)
+                conjuncts += ExprWithTypeConstraint(
+                    constraint = mkAnd(lhsType.boolTypeExpr, rhsType.boolTypeExpr),
+                    expr = onBool(lhs.extractBool(scope), rhs.extractBool(scope), scope)?.asExpr(bottomValueSort) ?: return null
+                )
 
-        override fun TsContext.onFp(
-            lhs: UExpr<KFp64Sort>,
-            rhs: UExpr<KFp64Sort>,
-            scope: TsStepScope,
-        ): UBoolExpr {
-            return mkFpEqualExpr(lhs, rhs)
-        }
+                // fake(bool) + fake(fp)
+                conjuncts += ExprWithTypeConstraint(
+                    constraint = mkAnd(lhsType.boolTypeExpr, rhsType.fpTypeExpr),
+                    expr = internalResolve(
+                        lhs.extractBool(scope),
+                        rhs.extractFp(scope),
+                        scope
+                    )?.asExpr(bottomValueSort) ?: return null
+                )
 
-        override fun TsContext.onRef(
-            lhs: UHeapRef,
-            rhs: UHeapRef,
-            scope: TsStepScope,
-        ): UBoolExpr {
-            return mkEq(lhs, rhs)
-        }
+                // fake(bool) + fake(ref)
+                conjuncts += ExprWithTypeConstraint(
+                    constraint = mkAnd(lhsType.boolTypeExpr, rhsType.refTypeExpr),
+                    expr = internalResolve(lhs.extractBool(scope), rhs.extractRef(scope), scope)?.asExpr(bottomValueSort) ?: return null
+                )
 
-        override fun TsContext.resolveFakeObject(
-            lhs: UExpr<*>,
-            rhs: UExpr<*>,
-            scope: TsStepScope,
-        ): UBoolExpr {
-            check(lhs.isFakeObject() || rhs.isFakeObject())
+                // fake(fp) + fake(bool)
+                conjuncts += ExprWithTypeConstraint(
+                    constraint = mkAnd(lhsType.fpTypeExpr, rhsType.boolTypeExpr),
+                    expr = internalResolve(
+                        lhs.extractFp(scope),
+                        rhs.extractBool(scope),
+                        scope
+                    )?.asExpr(bottomValueSort) ?: return null
+                )
 
-            val conjuncts = mutableListOf<ExprWithTypeConstraint<UBoolSort>>()
+                // fake(fp) + fake(fp)
+                conjuncts += ExprWithTypeConstraint(
+                    constraint = mkAnd(lhsType.fpTypeExpr, rhsType.fpTypeExpr),
+                    expr = onFp(lhs.extractFp(scope), rhs.extractFp(scope), scope)?.asExpr(bottomValueSort) ?: return null
```

Comment:
Consider placing arguments on separate lines, similar to above. Avoid long lines and inconsistent code.